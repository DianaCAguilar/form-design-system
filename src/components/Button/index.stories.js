import React from 'react';
import State from 'util/storybook-docs/State';
import { StoryWrapper, StoryItem } from 'util/storybook-docs/StoryLayout';
import StarFilledIcon from 'lib/icons/react/StarFilledIcon';
import WorldIcon from 'lib/icons/react/WorldIcon';
import Button, { THEMES, SIZES } from '.';

const Template = (args) => <Button label="Button" {...args} />;

export const Primary = Template.bind({});

export const Themes = () => (
  <StoryWrapper>
    {THEMES.map((theme, idx) => (
      <StoryItem key={idx}>
        <Button theme={theme} label={theme} />
      </StoryItem>
    ))}
  </StoryWrapper>
);

export const isDestructive = () => (
  <StoryWrapper>
    {THEMES.map((theme, idx) => (
      <StoryItem key={idx}>
        <Button isDestructive theme={theme} label={theme} />
      </StoryItem>
    ))}
  </StoryWrapper>
);

isDestructive.parameters = {
  docs: {
    description: {
      story:
        'Modifies the theme. Should only be used when the interaction involves data loss, and should also prompt the user to confirm the action.',
    },
  },
};

export const Inverted = () => (
  <div className="inverted bgColor--blue padding--all rounded--all">
    <StoryWrapper>
      <StoryItem>
        <Button theme="ghost" label="Button" />
      </StoryItem>
      <StoryItem>
        <Button theme="outlined" label="Button" />
      </StoryItem>
    </StoryWrapper>
  </div>
);

Inverted.parameters = {
  docs: {
    description: {
      story:
        'We currently support inverting `ghost` and `outlined` theme buttons. Simply surround the button(s) with an `.inverted` class.',
    },
  },
};

export const isActive = () => (
  <>
    <StoryWrapper>
      <State
        initialValue={{}}
        render={(value, setValue) => (
          <>
            {THEMES.map((theme, idx) => (
              <StoryItem key={idx}>
                <Button
                  theme={theme}
                  isActive={!value[idx]}
                  label={theme}
                  onClick={() => setValue({ ...value, [idx]: !value[idx] })}
                />
              </StoryItem>
            ))}
          </>
        )}
      />
      <State
        initialValue={{}}
        render={(value, setValue) => (
          <>
            {THEMES.map((theme, idx) => (
              <StoryItem key={idx}>
                <Button
                  theme={theme}
                  isActive={!value[idx]}
                  isDestructive
                  label={theme}
                  onClick={() => setValue({ ...value, [idx]: !value[idx] })}
                />
              </StoryItem>
            ))}
          </>
        )}
      />
    </StoryWrapper>
    <div className="inverted bgColor--blue padding--all rounded--all margin--top">
      <StoryWrapper>
        <State
          initialValue={{}}
          render={(value, setValue) => (
            <>
              {['ghost', 'outlined'].map((theme, idx) => (
                <StoryItem key={idx}>
                  <Button
                    theme={theme}
                    isActive={!value[idx]}
                    label={theme}
                    onClick={() => setValue({ ...value, [idx]: !value[idx] })}
                  />
                </StoryItem>
              ))}
            </>
          )}
        />
      </StoryWrapper>
    </div>
  </>
);

isActive.parameters = {
  docs: {
    description: {
      story:
        '`isActive` should be used on buttons to signify the button having some affect (applying a filter, causing a popover to be open). **Click to toggle on and off**.',
    },
  },
};

export const Sizes = () =>
  SIZES.map((size, idx) => (
    <div
      key={idx}
      className="display--inlineBlock margin--right--half margin--bottom--half"
    >
      <Button size={size} label="Button" />
    </div>
  ));

export const FullWidth = Template.bind({});
FullWidth.args = {
  isFullWidth: true,
};

export const Icons = () => (
  <StoryWrapper>
    <StoryItem>
      <Button IconLeft={WorldIcon} label="Button" />
    </StoryItem>
    <StoryItem>
      <Button IconRight={StarFilledIcon} label="Button" />
    </StoryItem>
  </StoryWrapper>
);

Icons.parameters = {
  docs: {
    description: {
      story:
        'ONLY pass in FDS Icon references to `IconLeft` or `IconRight`. Very important. (refer to code sample)',
    },
  },
};

export const Caret = () => (
  <StoryWrapper>
    <StoryItem>
      <Button theme="blue" hasCaret label="Caret" />
    </StoryItem>
    <StoryItem>
      <Button theme="blue" hasCaret IconRight={StarFilledIcon} label="Caret" />
    </StoryItem>
    <StoryItem>
      <Button theme="blue" hasCaret IconRight={StarFilledIcon} label="Caret" />
    </StoryItem>
  </StoryWrapper>
);

export const Loading = () => (
  <>
    <StoryWrapper>
      {THEMES.map((theme, idx) => (
        <StoryItem key={idx}>
          <Button theme={theme} isLoading label="Button" />
        </StoryItem>
      ))}
      {THEMES.map((theme, idx) => (
        <StoryItem key={idx}>
          <Button theme={theme} isDestructive isLoading label="Button" />
        </StoryItem>
      ))}
    </StoryWrapper>
    <div className="inverted bgColor--blue padding--all rounded--all margin--top">
      <StoryWrapper>
        <StoryItem>
          <Button theme="ghost" isLoading label="Button" />
        </StoryItem>
        <StoryItem>
          <Button theme="outlined" isLoading label="Button" />
        </StoryItem>
      </StoryWrapper>
    </div>
  </>
);

Loading.parameters = {
  docs: {
    description: {
      story:
        'Loading states should be shown whenever an action has the ability to take more than a couple of seconds of loading time',
    },
  },
};

export const Breakout = (args) => (
  <StoryWrapper>
    <StoryItem>
      <div
        className="alignChild--center--center padding--all--xl"
        style={{ border: '1px dashed #f09', position: 'relative' }}
      >
        <Button {...args} label="Button" isBreakoutLink />
      </div>
    </StoryItem>
  </StoryWrapper>
);

Breakout.parameters = {
  docs: {
    description: {
      story: 'Breakout buttons extend their clickable area to the next nearest parent.',
    },
  },
};

export const Misc = () => (
  <div style={{ width: '100px' }}>
    <Button label="Text can wrap" />
  </div>
);

export default {
  component: Button,
  title: 'components/Button',
  parameters: {
    docs: {
      attributes: ['rest'],
    },
  },
};
