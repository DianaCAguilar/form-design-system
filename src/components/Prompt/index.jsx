import React from 'react';
import PropTypes from 'prop-types';
import Dialog from 'components/Dialog';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';

const Prompt = (props) => (
  <Dialog
    content={props.desc ? <div id="a11y-dialog-desc">{props.desc}</div> : null}
    footerContent={
      <Flex justify="end" wrap>
        {props.secondaryButton && <FlexItem shrink>{props.secondaryButton}</FlexItem>}
        {props.primaryButton && <FlexItem shrink>{props.primaryButton}</FlexItem>}
      </Flex>
    }
    role="alertdialog"
    title={props.title}
    isOpen={props.isOpen}
    onDismiss={props.onDismiss}
    disablePortal={props.disablePortal}
  />
);

Prompt.propTypes = {
  /** Controls text in header */
  title: PropTypes.string,
  /** Controls string displayed in content area */
  desc: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
  /** Represents primary button (MUST be an FDS Button) */
  primaryButton: PropTypes.any.isRequired,
  /** Represents secondary button (MUST be an FDS Button) */
  secondaryButton: PropTypes.any.isRequired,
  /** Controls whether modal (and backdrop) are visible or not */
  isOpen: PropTypes.bool,
  /**
   * Callback that user can pass in, to be conditionally fired when
   * user attempts to close modal. When defined, the modal close button appears
   */
  onDismiss: PropTypes.func,
  /** Keeps the Prompt from being rendered in a portal and renders it locally instead. */
  disablePortal: PropTypes.bool,
};

export default Prompt;
