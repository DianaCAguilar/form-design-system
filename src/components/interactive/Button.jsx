import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import baseElement from '../../util/baseElement';

const Button = ({ theme, isLoading, iconPlacement, Icon, Link, children, ...rest }) => {
  const Element = baseElement({ href: rest.href, as: Link });
  return (
    <Element
      className={cx('fdsButton', {
        'fdsButton--contained-blue': theme === 'contained--blue',
        'fdsButton--contained-red': theme === 'contained--red',
        'fdsButton--loading': isLoading,
      })}
      {...rest}
    >
      {Icon && iconPlacement === 'left' && (
        <div className="margin--right--half fdsButton-icon">
          <Icon size="xs" />
        </div>
      )}
      <span className={cx({ 'fdsButton--hiddenLabel': isLoading })}>{children}</span>
      {Icon && iconPlacement === 'right' && (
        <div className="margin--left--half fdsButton-icon">
          <Icon size="xs" />
        </div>
      )}
    </Element>
  );
};

Button.defaultProps = {
  theme: 'primary--blue',
  iconPlacement: 'right',
};

Button.propTypes = {
  /**
   * Pass **only** react-router `Link` here. You may **not**
   * pass anything else here: SFC, Class Component, etc (even
   * if they use react-router `Link` underneath the hood).
   */
  Link: PropTypes.func,
  /**
   * Controls whether loading spinner displays. Button text and icons are set to visibility hidden
   * to preserve the space set, whilst hiding them
   */
  isLoading: PropTypes.bool,
  /** Controls which side the `Icon` renders on, assuming you pass it */
  iconPlacement: PropTypes.oneOf(['left', 'right']),
  /** Used to render a FDS Icon (should only be used for FDS Icons) */
  Icon: PropTypes.func,
  /** Used to control the display and theme of the button */
  theme: PropTypes.oneOf(['container--blue', 'container--red']),
  /** Contents inside the button */
  children: PropTypes.node,
};

export default Button;
