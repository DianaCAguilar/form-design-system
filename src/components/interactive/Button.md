# Button

Displays a standard Button.
