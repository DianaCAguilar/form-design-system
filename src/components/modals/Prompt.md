# Prompt

A confirm modal that displays a simple message, with 2 buttons to either confirm the action, or deny it (does not show close icon, and
can't be dismissed by pressing ESC or clicking backdrop)