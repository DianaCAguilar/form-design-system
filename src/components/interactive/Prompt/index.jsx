import React from 'react';
import PropTypes from 'prop-types';
import Dialog from '../Dialog';
import Button from '../Button';
import Flex from '../../layout/Flex';
import FlexItem from '../../layout/FlexItem';

const Reusable = (props) => (
  <div className="">
    <Flex justify="end" wrap>
      {props.secondaryButton && (
        <FlexItem shrink>
          <Button {...props.secondaryButton} />
        </FlexItem>
      )}
      <FlexItem shrink>
        <Button {...props.primaryButton} />
      </FlexItem>
    </Flex>
  </div>
);

Reusable.propTypes = {
  primaryButton: PropTypes.any.isRequired,
  secondaryButton: PropTypes.any,
};

const Prompt = (props) => (
  <Dialog
    footerContent={props.buttonPosition === 'footer' && <Reusable {...props} />}
    content={
      <React.Fragment>
        <div>{props.content}</div>
        {props.buttonPosition === 'content' && (
          <div className="margin--top--double">
            <Reusable {...props} />
          </div>
        )}
      </React.Fragment>
    }
    role="alertdialog"
    canDismiss={false}
    title={props.title}
    isOpen={props.isOpen}
    onDismiss={props.onDismiss}
  />
);

Prompt.defaultProps = {
  buttonPosition: 'content',
};

Prompt.propTypes = {
  title: PropTypes.string,
  content: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
  primaryButton: PropTypes.any.isRequired,
  secondaryButton: PropTypes.any,
  isOpen: PropTypes.bool,
  onDismiss: PropTypes.func,
  buttonPosition: PropTypes.oneOf(['content', 'footer']),
};

export default Prompt;
