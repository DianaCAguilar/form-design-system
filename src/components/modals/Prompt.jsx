import React from 'react';
import PropTypes from 'prop-types';
import Dialog from './Dialog';
import Button from '../interactive/Button';
import Flex from '../layout/Flex';
import FlexItem from '../layout/FlexItem';

const Prompt = (props) => (
  <Dialog
    content={
      <React.Fragment>
        <div id="a11y-dialog-desc">{props.desc}</div>
        <div className="margin--top--double">
          <div className="">
            <Flex justify="end" wrap>
              {props.secondaryButton && (
                <FlexItem shrink>
                  <Button {...props.secondaryButton} />
                </FlexItem>
              )}
              <FlexItem shrink>
                <Button {...props.primaryButton} />
              </FlexItem>
            </Flex>
          </div>
        </div>
      </React.Fragment>
    }
    role="alertdialog"
    canDismiss={false}
    title={props.title}
    isOpen={props.isOpen}
    onDismiss={props.onDismiss}
  />
);

Prompt.defaultProps = {
  buttonPosition: 'content',
};

Prompt.propTypes = {
  title: PropTypes.string,
  desc: PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
  primaryButton: PropTypes.any.isRequired,
  secondaryButton: PropTypes.any.isRequired,
  isOpen: PropTypes.bool,
  onDismiss: PropTypes.func,
  buttonPosition: PropTypes.oneOf(['content', 'footer']),
};

export default Prompt;
