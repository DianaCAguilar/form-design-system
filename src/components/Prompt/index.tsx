import React, { useEffect, useRef, useState } from 'react';
import cc from 'classcat';
import Flex from 'components/Flex';
import { DenyIcon } from 'icons/react';
import { IconButton } from 'components';
import FlexItem from 'components/FlexItem';
import { ButtonProps } from 'components/Button';
import * as AlertDialog from '@radix-ui/react-alert-dialog';
import rafSchd from 'raf-schd';

export const isElementOverflowing = ({
  current,
}: React.MutableRefObject<HTMLDivElement | null>): boolean => {
  // Checking for current first is safer just in case,
  // and will also fail gracefully in tests
  if (current) {
    return current.scrollHeight > current.clientHeight;
  }
  return false;
};

export interface PromptProps {
  /** Controls text in header */
  title?: string;
  /** Controls string displayed in content area */
  desc?: React.ReactNode | string;
  /** Represents primary button (MUST be an FDS Button) */
  primaryButton: React.ReactElement<ButtonProps>;
  /** Represents secondary button (MUST be an FDS Button) */
  secondaryButton: React.ReactElement<ButtonProps>;
  /** Controls whether modal (and backdrop) are visible or not */
  isOpen?: boolean;
  /**
   * Callback that user can pass in, to be conditionally fired when
   * user attempts to close modal. When defined, the modal close button appears
   */
  onDismiss?: () => void;
  /** Keeps the Prompt from being rendered in a portal and renders it locally instead. */
  disablePortal?: boolean;
}

const Prompt = ({
  title,
  desc,
  primaryButton,
  secondaryButton,
  isOpen,
  onDismiss,
}: PromptProps): JSX.Element => {
  const contentEl = useRef<HTMLDivElement>(null);
  const [isOverflowing, setIsOverflowing] = useState(false);
  const schdSetIsOverflowing = rafSchd(setIsOverflowing);

  const handleResize = () => {
    schdSetIsOverflowing(isElementOverflowing(contentEl));
  };

  useEffect(() => {
    handleResize(); // needs to fire once immediately on mount
    window.addEventListener('resize', handleResize);
    return () => {
      window.removeEventListener('resize', handleResize);
    };
  }, []);

  return (
    <AlertDialog.Root open={isOpen} onOpenChange={onDismiss}>
      <AlertDialog.Overlay className="dialog-overlay"></AlertDialog.Overlay>
      <div className="dialog-zIndex dialog-wrapper">
        <AlertDialog.Content
          // onPointerDownOutside={(event) => event.preventDefault()}
          asChild
          className="ease-in-out"
        >
          <div
            className="dialog elevation--3 border--focus--noTransition"
            role={'alertdialog'}
            aria-labelledby={title && 'a11y-dialog-title'}
            aria-describedby="a11y-dialog-desc"
            tabIndex={-1}
            aria-modal="true"
            style={{
              maxWidth: '500px',
              maxHeight: '80vh',
            }}
            data-testid="dialog"
          >
            {(title || onDismiss) && (
              <>
                <div className="dialog-header">
                  <div className="bgColor--white padding--all border--bottom">
                    <div>
                      <div className="padding--right--xl type--head4">
                        {title ? <span id="a11y-dialog-title">{title}</span> : '\u00A0'}{' '}
                        {/* There always needs to be something (even a space) in the header for display reasons */}
                      </div>
                    </div>
                    {onDismiss && (
                      <AlertDialog.Trigger asChild>
                        <div className="dialog-icon">
                          <IconButton Icon={DenyIcon} aria-label="Close" label="Close" />
                        </div>
                      </AlertDialog.Trigger>
                    )}
                  </div>
                </div>
              </>
            )}
            {desc && (
              <div className="dialog-content" ref={contentEl}>
                <div className="padding--all bgColor--white">
                  {desc ? <div id="a11y-dialog-desc">{desc}</div> : null}
                </div>
              </div>
            )}
            {
              <div className="dialog-footer">
                <div
                  className={cc([
                    {
                      'border--top': isOverflowing,
                    },
                    'bgColor--white',
                    'padding--all',
                  ])}
                >
                  <Flex justify="end">
                    {secondaryButton && <FlexItem shrink>{secondaryButton}</FlexItem>}
                    {primaryButton && <FlexItem shrink>{primaryButton}</FlexItem>}
                  </Flex>
                </div>
              </div>
            }
          </div>
        </AlertDialog.Content>
      </div>
    </AlertDialog.Root>
  );
};

export default Prompt;
