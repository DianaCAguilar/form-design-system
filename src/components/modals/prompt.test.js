import React from 'react';
import { shallow } from 'enzyme';
import Prompt from './Prompt'
import Button from '../interactive/Button'

describe('Prompt component', () => {

  // Dialog component uses a library that uses window.scroll underneath the hood
  // so we just need to add this here to avoid errors
  window.scroll = () => { }

  it('matches snapshot (default props)', () => {
    const component = shallow(<Prompt primaryButton={<Button>foo</Button>} secondaryButton={<Button>bar</Button>} />);
    expect(component).toMatchSnapshot();
    const component2 = shallow(<Prompt isOpen={true} primaryButton={<Button>foo</Button>} secondaryButton={<Button>bar</Button>} />);
    expect(component2).toMatchSnapshot();
  });

  it('matches snapshot (set all props)', () => {
    const component = shallow(
      <Prompt 
        primaryButton={<Button>foo</Button>} 
        secondaryButton={<Button>bar</Button>} 
        title="title"
        desc="desc"
        isOpen={true}
      />
    );
    expect(component).toMatchSnapshot();
  });

})