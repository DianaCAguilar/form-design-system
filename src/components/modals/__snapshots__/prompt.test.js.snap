// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prompt component matches snapshot (default props) 1`] = `
<Dialog
  content={
    <div
      id="a11y-dialog-desc"
    />
  }
  footerContent={
    <div
      className="prompt-footer"
    >
      <div>
        <Button
          label="bar"
        />
      </div>
      <div>
        <Button
          label="foo"
        />
      </div>
    </div>
  }
  role="alertdialog"
/>
`;

exports[`Prompt component matches snapshot (default props) 2`] = `
<Dialog
  content={
    <div
      id="a11y-dialog-desc"
    />
  }
  footerContent={
    <div
      className="prompt-footer"
    >
      <div>
        <Button
          label="bar"
        />
      </div>
      <div>
        <Button
          label="foo"
        />
      </div>
    </div>
  }
  isOpen={true}
  role="alertdialog"
/>
`;

exports[`Prompt component matches snapshot (set all props) 1`] = `
<Dialog
  content={
    <div
      id="a11y-dialog-desc"
    >
      desc
    </div>
  }
  footerContent={
    <div
      className="prompt-footer"
    >
      <div>
        <Button
          label="bar"
        />
      </div>
      <div>
        <Button
          label="foo"
        />
      </div>
    </div>
  }
  isOpen={true}
  role="alertdialog"
  title="title"
/>
`;
